### ISSUE TITLE
Structure and File-names not generally compatible

### ISSUE THREAD

**Issue by BrendanBenshoof** [whyrusleeping/gx#14]
If you are only planning on using gx for node and go, then it works fine.

however other languages have different requirements for folder structures and limits to file name (python stands out as an example).

If you want it to be a "general any language" package manager, you need to find an elegant way to deal with this (language specific plugins?)

**Comment by whyrusleeping** [comment]
could you detail how I would need to let things be structured for python to work? (i only ever script in python, never done an actual 'project' in it)

**Comment by BrendanBenshoof** [comment]
Relative imports for python are iffy to start with. I don't advise on planning to use them that way.

The way to do it for python is to put packages somewhere on the python path (varies from system to system)

the "right" way to deal with python would be to require the user to add a dedicated "vendor" folder for the entire machine, then put packages there. 

For python you also cannot have a package name that would not work as a valid identifier in python. (no "."s or similar symbols, must not start with a number) Ironically a multihash works well. Your current name-version.num scheme cannot be imported.

When figuring out how to do this on my own, I named the package as a multihash and did:

import Qmblahblhablah as packagename

but I am not sure if people would tolerate that.

**Comment by whyrusleeping** [comment]
interesting... how does python normally do vendoring?

**Comment by BrendanBenshoof** [comment]
It doesn't. I essentially had to re-write every import in every module I've vendor-ed into urdht.

Essentially, you can do relative imports of modules inside the current module, but all later imports have to be relative to the "topmost" module rather than the current one. I like python, but that is nuts.

**Comment by whyrusleeping** [comment]
alright, how could gx make it easy for you?

**Comment by whyrusleeping** [comment]
`gx build` could set a custom python-path for you, or something maybe

**Comment by jbenet** [comment]
most "real world" (devops) use of python is strictly jails with venv. it is "absolute" in terms of the python path, but relative in terms of the OS.

i think language plugins (rules?) is going to be the right thing to do. will not be able to capture every lang.

**Comment by whyrusleeping** [comment]
yeah, my thought is having language specific plugins to handle things

**Comment by BrendanBenshoof** [comment]
The hard part is maintaining "composibility" such that when you recursively get the requirements for each package, they just work too.

**Comment by noffle** [comment]
Having a library for `gx` that separates these concerns could help here: #15.

**Comment by parkan** [comment]
I recently ended up building a similar feature for mediachain's experimental [translators](https://github.com/mediachain/mediachain/blob/master/docs/translators.md), which ended up doing a lot of ugly `sys.path` manipulation, namespace juggling and `__import__`ing (we needed dynamic loading)

https://github.com/mediachain/mediachain-client/blob/master/mediachain/translation/lookup.py#L37

One issue I encountered was namespace shadowing; any `.pyc` on the path (python maddeningly sets `sys.path[0]` to cwd even if you're in a venv) that provide any part of the package prefix will shadow any branches in `site-packages`, e.g. if you have `foo.bar` in any directory on the path preceding the venv vendored dir, `foo.qux` from there will never be found.

It would be great to have a general purpose solution to this!

I wonder if a cleaner way to do this is to symlink (under the real package name instead of versioned/content-addressed path)?
